### PR TITLE
created button for library export

### PR DESCRIPTION
**Pull request by wgu-edwin**
button checks auth roles to make sure only users with access can export

### CODE REVIEW

File: ui/src/app/navigation/commoncontrols.component.html [hunk 1]
```diff
@@ -65,6 +65,12 @@
           </svg>
           <span class="m-button-x-text">Batch Import RSD<span class="t-type-lowercase">s</span></span>
         </a>
+        <a class="m-button" routerLink="/skills/export" [class.disabled-content]="!canExportLibrary">
```
Comment (JohnKallies) [inline ui/src/app/navigation/commoncontrols.component.html:68]: Let's make is visible / not visible, rather than enabled / disabled?

**Review by JohnKallies — COMMENTED** [review]
This is a good start. It needs a route and a server call still

File: .github/workflows/maven-release.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# This workflow will build maven releases
```
Comment (JohnKallies) [inline .github/workflows/maven-release.yml:1]: This change does not belong in this PR. We can work together on your git history if you need help applying just your feature changes to the develop branch.

File: ui/src/app/richskill/service/rich-skill.service.ts [hunk 3]
```diff
@@ -149,6 +149,30 @@ export class RichSkillService extends AbstractService {
       }))
   }
 
+  libraryExport(
+    publishStatus: Set<PublishStatus>,
```
Comment (JohnKallies) [inline ui/src/app/richskill/service/rich-skill.service.ts:153]: Why are we providing search and paging params to a request to dump the entire library?

File: ui/src/app/richskill/service/rich-skill.service.ts [hunk 4]
```diff
@@ -149,6 +149,30 @@ export class RichSkillService extends AbstractService {
       }))
   }
 
+  libraryExport(
```
Comment (JohnKallies) [inline ui/src/app/richskill/service/rich-skill.service.ts:152]: This needs corresponding test coverage.

File: ui/src/app/richskill/service/rich-skill.service.ts (hunk 3 above)
Comment (wgu-edwin) [reply ui/src/app/richskill/service/rich-skill.service.ts:153]: I asked this too, but Jose mentioned he created the API route with those params, so i'm just passing what he requested.

File: ui/src/app/navigation/commoncontrols.component.ts [hunk 5]
```diff
@@ -14,10 +19,26 @@ export class CommoncontrolsComponent extends AbstractSearchComponent implements
   searchIcon = SvgHelper.path(SvgIcon.SEARCH)
   dismissIcon = SvgHelper.path(SvgIcon.DISMISS)
 
-  constructor(protected searchService: SearchService, protected route: ActivatedRoute, protected authService: AuthService) {
+  selectedFilters: Set<PublishStatus> = new Set([PublishStatus.Draft, PublishStatus.Published])
```
[the remark this answers is not in the document]
Comment (wgu-edwin) [reply ui/src/app/navigation/commoncontrols.component.ts:22]: @JohnKallies  I put the logic in common controls because the button lives inside the template with the same name. We could put the logic in a parent component or the abstract search component that it extends from?

File: ui/src/app/navigation/commoncontrols.component.ts [hunk 6]
```diff
@@ -14,10 +17,24 @@ export class CommoncontrolsComponent extends AbstractSearchComponent implements
   searchIcon = SvgHelper.path(SvgIcon.SEARCH)
   dismissIcon = SvgHelper.path(SvgIcon.DISMISS)
 
-  constructor(protected searchService: SearchService, protected route: ActivatedRoute, protected authService: AuthService) {
+  constructor(
```
Comment (JohnKallies) [inline ui/src/app/navigation/commoncontrols.component.ts:20]: This is legit, but weird. Please annotate why we put it here with a code comment.

File: ui/src/app/navigation/commoncontrols.component.html [hunk 7]
```diff
@@ -65,7 +65,12 @@
           </svg>
           <span class="m-button-x-text">Batch Import RSD<span class="t-type-lowercase">s</span></span>
         </a>
-
+        <button class="m-button" type="button" (click)=onDownloadCsv() *ngIf="canExportLibrary">
+          <svg class="m-button-x-icon t-icon">
+            <use xlink:href="/assets/images/svg-defs.svg#icon-download"></use>
+          </svg>
+          <span class="m-button-x-text">Export RSD Library<span class="t-type-lowercase"></span></span>
```
Comment (hikr17) [inline ui/src/app/navigation/commoncontrols.component.html:72]: Looks like a double closure </span></span> that mismatches the single <span> opening.

File: ui/src/app/navigation/commoncontrols.component.ts [hunk 8]
```diff
@@ -14,10 +17,30 @@ export class CommoncontrolsComponent extends AbstractSearchComponent implements
...
+    protected searchService: SearchService,
+    protected route: ActivatedRoute,
+    protected authService: AuthService,
+    protected richSkillService: RichSkillService,
+    @Inject(LOCALE_ID) protected locale: string) {
```
Comment (hikr17) [inline ui/src/app/navigation/commoncontrols.component.ts:25]: You're right.  Adding so many dependencies implies to me that the file wasn't intended to address the new code.  Perhaps the export code belongs somewhere else?  Could it be put in a LibraryExportComponent.ts?

Comment (hikr17) [reply ui/src/app/navigation/commoncontrols.component.ts:25]: Or if it just supports the rich-skill.service.ts, then perhaps the new code belongs there.  Just thinking about it is all.

File: ui/src/app/richskill/service/rich-skill.service.spec.ts [hunk 9]
```diff
@@ -306,6 +306,23 @@ describe("RichSkillService", () => {
...
 
+  it("libraryExport should return", () => {
+    RouterData.commands = []
+
+    // Act
+    const result$ = testService.libraryExport()
+
+    // Assert
+    result$.subscribe( (data: string) => {
+      expect(RouterData.commands).toEqual([]) // No Errors
```
Comment (hikr17) [inline ui/src/app/richskill/service/rich-skill.service.spec.ts:317]: It doesn't look like RouterData.commands ever gets assigned a new value.  Also, the subscription data parameter is never used.  Perhaps that's what was intended.

Comment (hikr17) [reply ui/src/app/richskill/service/rich-skill.service.spec.ts:317]: I don't recall right now, but since there is an expect() statement in the callback, you might run into occasional test failures.  If that happens, we may have to mark the test case as asynchronous.  There are >=2 ways to deal with that situation.

File: ui/src/app/navigation/libraryexport.component.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+  ngOnInit(): void {
+  }
+
+  onDownloadLibrary(): void {
+    this.richSkillService.libraryExport()
+      .subscribe((csv: string) => {
+        const blob = new Blob([csv], {type: "text/csv;charset=utf-8;"})
+        const date = formatDate(new Date(), "yyyy-MM-dd", this.locale)
+        FileSaver.saveAs(blob, `RSD Library - OSMT ${date}.csv`)
```
Comment (JohnKallies) [inline ui/src/app/navigation/libraryexport.component.ts:38]: @wgu-edwin, I'd like to see this covered by unit test as well. Something along the lines of

```
    spyOn(FileSaver, 'saveAs').and.stub();
```
Looking for saveAs to be called with a Blob of type CSV and a file name matching most of the template string in the code (startsWith and endsWith. Don't get hung up on the date, that's enough coverage to prevent accidental drift in the future).

File: ui/test/resource/mock-stubs.ts [hunk 11]
```diff
@@ -336,6 +336,11 @@ export class RichSkillServiceStub {
   ): Observable<PaginatedSkills> {
     return of(createMockPaginatedSkills())
   }
+
+  // noinspection JSUnusedGlobalSymbols,JSUnusedLocalSymbols
+  libraryExport(): Observable<string> {
+    return of(`x, y, z`)
```
Comment (hikr17) [inline ui/test/resource/mock-stubs.ts:342]: Does this do anything of value?  Did you intend to keep it?

Comment (wgu-edwin) [reply ui/test/resource/mock-stubs.ts:342]: I will probably change it to something that resembles csv data. The other tests that return csv had this as the return hence why i tried it to test if it works for now

File: ui/src/app/app-routing.module.ts [hunk 12]
```diff
@@ -21,6 +21,7 @@ import {PublishCollectionComponent} from "./collection/detail/publish-collection
 import {CollectionSkillSearchComponent} from "./collection/collection-skill-search.component"
 import {BatchImportComponent} from "./richskill/import/batch-import.component"
 import { ActionByRoles, ButtonAction } from "./auth/auth-roles"
+import {CommoncontrolsComponent} from "./navigation/commoncontrols.component"
```
Comment (JohnKallies) [inline ui/src/app/app-routing.module.ts:24]: Is this needed? If not, please remove.

File: ui/src/app/navigation/commoncontrols.component.ts [hunk 13]
```diff
@@ -14,7 +14,10 @@ export class CommoncontrolsComponent extends AbstractSearchComponent implements
   searchIcon = SvgHelper.path(SvgIcon.SEARCH)
   dismissIcon = SvgHelper.path(SvgIcon.DISMISS)
 
-  constructor(protected searchService: SearchService, protected route: ActivatedRoute, protected authService: AuthService) {
```
Comment (JohnKallies) [inline ui/src/app/navigation/commoncontrols.component.ts:17]: It doesn't look like you made any functional changes in this file. Can you please revert the whitespace changes?

File: ui/src/app/navigation/libraryexport.component.ts (hunk 10 above)
Comment (JohnKallies) [reply ui/src/app/navigation/libraryexport.component.ts:37]: Per our verbal, we are going to punt on fine-grained testing of the arguments for now, and upgrade to the latest Jasmine, which makes that easy.

File: ui/test/resource/mock-stubs.ts (hunk 11 above)
Comment (JohnKallies) [reply ui/test/resource/mock-stubs.ts:342]: @wgu-edwin, if this change isn't being used, let's remove it? I can't tell if it's providing data for a test or not.

File: ui/src/app/navigation/commoncontrols.component.html (hunk 7 above)
Comment (hikr17) [reply ui/src/app/navigation/commoncontrols.component.html]: My bad.

File: ui/src/app/app-routing.module.ts (hunk 12 above)
Comment (wgu-edwin) [reply ui/src/app/app-routing.module.ts]: I have removed the unused variable

File: ui/src/app/richskill/service/rich-skill.service.spec.ts (hunk 9 above)
Comment (wgu-edwin) [reply ui/src/app/richskill/service/rich-skill.service.spec.ts:318]: I have added the fakeAsync() so it doesn't fail.

**Review by JohnKallies — APPROVED** [review]
Thanks for tracking the details. To me, this matters.

**Comment by JohnKallies** [comment]
I see this UI work as complete, nicely done. We will merge this after the API code for export.

**Comment by JohnKallies** [comment]
Completed in #265
